app/routes/products.$productSlug/route.tsx

```diff
@@ -14,18 +15,55 @@ import { ProductOption } from '~/components/product-option/product-option';
 import { UnsafeRichText } from '~/components/rich-text/rich-text';
 import { ROUTES } from '~/router/config';
 import {
+    getChoiceValue,
     getErrorMessage,
+    getMatchingVariants,
+    getMedia,
     getPriceData,
-    selectedChoicesToVariantChoices,
     getSelectedVariant,
     getSKU,
     getUrlOriginWithPath,
     isOutOfStock,
-    getMedia,
+    selectedChoicesToVariantChoices,
 } from '~/utils';
-import { AddToCartOptions, EcomApiErrorCodes } from '~/api/types';
 import styles from './product-details.module.scss';
 
+function isChoiceVisible(
+    choice: products.Choice,
+    option: products.ProductOption,
+    selectedChoices: Record<string, products.Choice | undefined>,
+    product: Product | SerializeFrom<Product>
+): boolean {
+    if (!option.name || !option.optionType) {
+        return false;
+    }
+
+    // Get variants matching all other selected choices except the current option
+    const matchingVariants = getMatchingVariants(product, {
+        ...selectedChoices,
+        [option.name]: undefined,
+    });
+
+    const choiceValue = getChoiceValue(option.optionType, choice);
+    return matchingVariants.some(
+        (variant) => variant.variant?.visible && variant.choices?.[option.name!] === choiceValue
+    );
+}
+
+function getActualOptions(
```

Comment:
maybe move this to product utils and name it just `getProductOptions` with the comment what it does?